Ports/build_all.sh

```diff
@@ -1,89 +1,109 @@
 #!/usr/bin/env bash
 
-clean=false
-verbose=false
+some_failed='false'
+action='build'
+verbose='false'
+failfast='false'
 
-case "$1" in
-    clean)
-        clean=true
-        ;;
-    verbose)
-        verbose=true
-        ;;
-    *)
-        ;;
-esac
-
-case "$2" in
-    clean)
-        clean=true
-        ;;
-    verbose)
-        verbose=true
-        ;;
-    *)
-        ;;
-esac
+for arg in "$@"; do
+    case "$arg" in
+        clean*)
+            action="$arg"
+            ;;
+        verbose)
+            verbose=true
+            ;;
+        failfast)
+            failfast=true
+            ;;
+    esac
+done
 
 some_failed=false
-built_ports=""
+processed_ports=()
 
-for file in *; do
-    if [ -d $file ]; then
-        pushd $file > /dev/null
-            port=$(basename $file)
-            port_built=0
-            for built_port in $built_ports; do
-                if [ "$built_port" = "$port" ]; then
-                    port_built=1
-                    break
-                fi
-            done
-            if [ $port_built -eq 1 ]; then
-                echo "Built $port."
-                popd > /dev/null
-                continue
-            fi
-            if ! [ -f package.sh ]; then
-                echo "ERROR: Skipping $port because its package.sh script is missing."
-                popd > /dev/null
-                continue
-            fi
+log_success() {
+    echo -e "\033[1;32m[#]\033[0m $1"
+}
+log_warn() {
+    echo -e "\033[1;33m[!]\033[0m $1"
+}
+log_info() {
+    echo -e "\033[1;36m[*]\033[0m $1"
+}
+log_process() {
+    echo -e "\033[1m[~]\033[0m $1"
+}
+log_error() {
+    echo -e "\033[1;31m[x]\033[0m $1"
+}
+
+do_build_port() {
+    log_process "Building $port_name"
+    if $verbose; then
+        ./package.sh
+    else
+        ./package.sh &> /dev/null
+    fi
+}
 
-            if [ "$clean" == true ]; then
-                if [ "$verbose" == true ]; then
-                    ./package.sh clean_all
-                else
-                    ./package.sh clean_all > /dev/null 2>&1
+do_clean_port() {
+    log_process "Cleaning $port_name"
+    if $verbose; then
+        ./package.sh "$1"
+    else
+        ./package.sh "$1" &> /dev/null
+    fi
+}
+
+# It is required to simply implement cd, instead of pushd-popd
```

Comment:
Last time you said to remove realpath and use dirname directly. But I am using `cd` only once. If realpath is not implemented then lets say it first changed directory to `./Ports/aclock` and now new loop body executes, it will try to cd into `./Ports/aclock/./Ports/arp` which is not a valid directory.

So with realpath, we don't have to worry about cd back into main Ports directory. As in first time its `/mnt/Projects/serenity/Ports/aclock` and then `/mnt/Projects/serenity/Ports/arp`. Both are valid directories and `cd` statement is used only once which could avoid potential problem of missing cd.

Reply:
I don't think I said to not use `realpath`, I meant that we don't need a separate variable just to store its result (back then you were making another variable just to append `/Ports` to the string).

In any case, that comment doesn't make much sense like this, so I'd just omit it. Also, please note the other comment that I left about returning to the original directory afterwards.

Reply:
Is it clear to you now what that comment mean? I will remove it for now.

Reply:
Yes, it's clear to me now. However, it will most likely confuse anyone who isn't aware of this conversation, so let's remove it.

Reply:
Removed. Please check it.